File: Code/Python/josephus_problem.py
# Josephus Problem Statement:
# We are given n people, standing in a circle.
# In each step, k-1 people are skipped and the kth person is executed(killed)
# The process stops when only one person remains. It is required to find the position of last person.

# strategy:
# josephus(n,k) can be calculated if we know the solution of josephus(n-1,k)
# Therefore, this will be solved using recursion.


def josephus(n, k):
    # only one person will be left, which is the answer
    if n == 1:
        return 1
    else:

        # The recursive call
        # josephus(n-1,k) considers the
        # original position k%n + 1 as position 1
        # therefore the returned position
        # is adjusted.

        return (josephus(n - 1, k) + k - 1) % n + 1


n = int(input())
k = int(input())
print(josephus(n, k))



# Test case 1:
#    Input:
#    6 3
#    Output:
#    1

# Test case 2:
#    Input:
#    15 4
#    Output:
#    13

# Time Complexity: O(n)
# Space Complexity: O(n)

File: Code/C++/josephus_problem.cpp
/*

This is a theoretical computer science problem.
There are n people in a circle waiting to be executed.
Starting from any position, a person starts executing the k'th person ahead of him, 
then the person after the one who is executed repeats the same task, 
as there will be only one person surviving, we have to find the person who will remain alive.

*/

#include <iostream>
#include <vector>
using namespace std;

void josephus(vector<int> &v, int k, int idx, int &ans)
{
    if (v.size() == 1)
    { // only one person will be left, which is the answer
        ans = v[0];
        return;
    }

    idx = (idx + k - 1) % v.size();
    v.erase(v.begin() + idx);
    josephus(v, k, idx, ans);
}

int main()
{
    // use vector, if persons are not listed from 1 to n
    vector<int> lst;
    int n, k, ans = 0;
    cin >> n >> k;
    lst.reserve(n);
    for (int i = 0; i < n; i++)
    {
        lst.push_back(i + 1);
    }
    // lst = {1,2,3,4,..,n}
    josephus(lst, k, 0, ans);
    cout << ans << endl;
    return 0;
}

/*

Test Case:

Input: 3 2
Output: 3

Input: 4 6
Output: 3

Input: 32 32
Output: 27

Time Complexity: O(n)
Space Complexity: O(n)

*/
